src/vm/docs/mod.rs

```diff
@@ -737,6 +747,115 @@ definition (i.e., you cannot put a define statement in the middle of a function
+    signature: "(mint-token! token-name amount recipient)",
+    description: "`mint-token!` is used to increase the token balance for the `recipient` principal for a token
+type defined using `define-token`. The increased token balance is _not_ transfered from another principal, but
+rather minted.",
+    example: "
+(define-token stackaroo)
+(mint-token! stackaroo 100 tx-sender)
+"
+};
+
+const MINT_ASSET: SpecialAPI = SpecialAPI {
+    name: "mint-asset!",
+    input_type: "AssetName, A, principal",
+    output_type: "(response bool int)",
+    signature: "(mint-asset! asset-class asset-identifier recipient)",
+    description: "`mint-asset!` is used to instantiate an asset and set that asset's owner to the `recipient` principal.
+The asset must have been defined using `define-asset`, and the supplied `asset-identifier` must be of the same type specified in
+that definition.
+
+If an asset identified by `asset-identifier` _already exists_, this function will return an error with the following error code:
+
+`(err 1)`
+",
+    example: "
+(define-asset stackaroo (buff 40))
+(mint-token! stackaroo \"Roo\" tx-sender)
+"
+};
+
+const GET_OWNER: SpecialAPI = SpecialAPI {
+    name: "get-owner",
+    input_type: "AssetName, A",
+    output_type: "(optional principal)",
+    signature: "(get-owner asset-class asset-identifier)",
+    description: "`get-owner` returns the owner of an asset, identified by `asset-identifier`, or `none` if the asset does not exist.
+The asset type must have been defined using `define-asset`, and the supplied `asset-identifier` must be of the same type specified in
+that definition.",
+    example: "
+(define-asset stackaroo (buff 40))
+(get-owner stackaroo \"Roo\")
+"
+};
+
+
+const GET_BALANCE: SpecialAPI = SpecialAPI {
+    name: "get-balance",
+    input_type: "TokenName, principal",
+    output_type: "int",
+    signature: "(get-balance token-name principal)",
+    description: "`get-balance` returns `token-name` balance of the principal `principal`.
+The token type must have been defined using `define-token`.",
+    example: "
+(define-token stackaroos)
+(get-balance stackaroos tx-sender)
+"
+};
+
+const TOKEN_TRANSFER: SpecialAPI = SpecialAPI {
+    name: "transfer-token!",
+    input_type: "TokenName, int, principal, principal",
+    output_type: "(response bool int)",
+    signature: "(transfer-token! token-name amount sender recipient)",
+    description: "`transfer-token!` is used to increase the token balance for the `recipient` principal for a token
+type defined using `define-token` by debiting the `sender` principal.
+
+This function returns (ok true) if the transfer is successful. In the event of an unsuccessful transfer it returns
+one of the following error codes:
+
+`(err 1)` -- `sender` does not have enough balance to transfer
+`(err 2)` -- `sender` and `recipient` are the same principal
+`(err 3)` -- amount to send is non-positive
+",
+    example: "
+(define-token stackaroo)
```

Comment:
I think this is a good idea since it would make it possible to reject transactions if they over-inflate the supply without having to rely on developers to do the check.  Also, all tokens are implicitly capped at a maximum supply of `2**127 - 1` due to our use of `i128` to encode token quantities.

Reply:
I disagree -- setting a `total_supply` forces us to make more opinionated choices about how token supplies work, and this is a check that contract authors can perform themselves very easily.

Reply:
How messy would an optional 2nd param for `total_supply` be? 
Without it and the associated checks, this increases the code audit burden for token smart contracts authors who claim that there is a finite supply.

Reply:
@kantai I agree with @zone117x on this one -- there are a plethora of on-chain tokens that will have a total supply, and making it hard for them to screw up enforcing a total supply would be beneficial to everyone.

But regardless of how a total supply is enforced, this discussion nevertheless raises the related question of how we intend to support tokens that do _not_ have a total supply, if at all.  We currently don't support the notion of "infinite" tokens.  I personally don't think we should try to do so, since trying to do so introduces the problem of keeping the number of bytes required to encode absurdly large numbers under control.  But if we don't support "infinite" tokens, then every token already implicitly has a "total_supply" of 2**127 - 1.  Given this, I think we're already forcing a choice about token supplies work -- it's just that we're doing so implicitly if we don't offer a `total_supply` field.

Reply:
We do support infinite supplies, though. It's only individual address balances that must be 2*127 - 1.

Implementing a finite supply check is very straight-forward:

```scheme
(define total-supply 10)
(define-var current-supply int 0)
(define (protected-mint (for principal) (amount int))
    (if (> (+ amount current-supply) total-supply)
        (err 0)
        (let (mint-result (mint! token amount for))
              (if (is-ok? mint-result)
                  (begin (set-var! current-supply (+ current-supply amount))
                             mint-result)
                  mint-result))))
```

Reply:
Ultimately, I'm okay with adding an optional parameter here, I just think we should be conscious of adding too much syntactic sugar (which, probably isn't the case here -- but we should come to some kind of conclusion about when new syntax is appropriate versus not).

Reply:
D'oh! you're right -- there's only a limit on a single account.

I think in order to answer the question of whether or not to enforce `total_supply` we need to see if there's been a history of smart contract bugs that have lead to unintentional inflation.  I'll be convinced to _not_ implement `total_supply` if (1) there are plenty of smart contracts in the wild that enforce a total supply on their own and (2) few/none of them have had bugs from doing it.

Reply:
https://github.com/sec-bit/awesome-buggy-erc20-tokens/blob/master/bad_tokens.top.json
ctrl+F `totalsupply-overflow`

Reply:
I'm a fan of avoiding syntactic sugar if it leads to ambiguity.  We could have `define-capped-token` as well as `define-token`, for example.